### PR TITLE
feat(automation) : Add line numbers to keyword and copyright scanners

### PR DESCRIPTION
**Pull request by rajuljha**
## Description

Add line numbers for the copyright and keyword scanners for diff scans

### Changes

- Introduced a new class FormatResult to handle adding line numbers to the output of the scanners
- Modified existing print results functionality to include the line number results.
- Add new functions in main script `fossologyscanner.py` to format the results of scanners
- Add a new class `ScanResultList`, which inherits from `ScanResult` to hold full content from the scan results
- Refactored the scanner module functions to include 'whole` flags for spitting complete info.

## How to test

- Build the docker image from the `Dockerfile.ci` located at `utils/automation` directory.
- Set up a Github Actions workflow in any of your repositories with this [example](https://github.com/fossology/fossology/blob/master/utils/automation/.github-workflow.yml)
- Set up a Github Fine Grained Token ([link](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#creating-a-fine-grained-personal-access-token))
- Set up a environment variable file and include the following Environment variables: 
```
GITHUB_ACTIONS=true
GITHUB_TOKEN=<your_github_token>
GITHUB_REPOSITORY=<your_gihtub_repo>
GITHUB_REPO_OWNER=<your_github_username>
GITHUB_REPO_URL=<youor_repo_url>
GITHUB_PULL_REQUEST=<your_pull_request_id>
```
- Run the docker container built with passing this environment variable file as a parameter
- Go to `/bin` directory and run the `fossologyscanner.py` script.

### Depends on 
This PR depends on #2785

### CODE REVIEW

**Comment by rajuljha** [comment]
> ## Description
> 
> Add line numbers for the copyright and keyword scanners for diff scans
> ### Changes
> 
>     * Introduced a new class FormatResult to handle adding line numbers to the output of the scanners
> 
>     * Modified existing print results functionality to include the line number results.
> 
>     * Add new functions in main script `fossologyscanner.py` to format the results of scanners
> 
>     * Add two new methods to the Scanner class
> 
> 
> ## How to test
> 
>     * Build the docker image from the `Dockerfile.ci` located at `utils/automation` directory.
> 
>     * Set up a Github Actions workflow in any of your repositories with this [example](https://github.com/fossology/fossology/blob/master/utils/automation/.github-workflow.yml)
> 
>     * Set up a Github Fine Grained Token ([link](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#creating-a-fine-grained-personal-access-token))
> 
>     * Set up a environment variable file and include the following Environment variables:
> 
> 
> ```
> GITHUB_ACTIONS=true
> GITHUB_TOKEN=<your_github_token>
> GITHUB_REPOSITORY=<your_gihtub_repo>
> GITHUB_REPO_OWNER=<your_github_username>
> GITHUB_REPO_URL=<youor_repo_url>
> GITHUB_PULL_REQUEST=<your_pull_request_id>
> ```
> 
>     * Run the docker container built with passing this environment variable file as a parameter
> 
>     * Go to `/bin` directory and run the `fossologyscanner.py` script.

CC @GMishx @shaheemazmalmmd

File: utils/automation/fossologyscanner.py [hunk 1]
```diff
@@ -82,13 +83,14 @@ def get_allow_list() -> dict:
   return data
 
 
-def print_results(name: str, failed_results: List[ScanResult],
+def print_results(name: str, failed_results: List[ScanResult],scan_results_with_line_number:List[dict],
```
Comment (GMishx) [inline utils/automation/fossologyscanner.py:86]: Please keep the line length limited to 80 characters :-). Please put the new parameter on new line.

File: utils/automation/fossologyscanner.py [hunk 2]
```diff
@@ -140,38 +152,92 @@ def print_log_message(filename: str,
...
+
+  :param: scanner : Scanner Scanner object
+  :param: format_results : FormatResult FormatResult object
+  :return: list of dicts with key as word and value as list of line numbers of the words
+  """
+  keyword_results = scanner.get_keyword_whole()
+  formatted_list_of_keyword_line_numbers = list()
+  for keyword_result in keyword_results:
+    list_of_scan_results = list(keyword_result.result)
+    words_to_search = [list_of_scan_results[i] for i in range(0,len(list_of_scan_results))]
```
Comment (GMishx) [inline utils/automation/fossologyscanner.py:167]: This line seems to be doing nothing. Can you please explain the need for this list comprehension?

File: utils/automation/fossologyscanner.py [hunk 3]
```diff
@@ -140,38 +152,92 @@ def print_log_message(filename: str,
...
+  for license_result in license_results:
+    list_of_scan_results = list(license_result.result)
+    words_to_search = [list_of_scan_results[i] for i in range(0,len(list_of_scan_results))]
+    words_with_line_numbers = format_results.find_word_line_numbers_without_bytes(license_result.path,words_to_search)
+    formatted_list_of_license_line_numbers.append(words_with_line_numbers)
+  return formatted_list_of_license_line_numbers
 
 def text_report(cli_options: CliOptions, result_dir: str, return_val: int,
-                scanner: Scanners) -> int:
+                scanner: Scanners, format_results = FormatResult) -> int:
```
Comment (GMishx) [inline utils/automation/fossologyscanner.py:207]: You meant?
```suggestion
                scanner: Scanners, format_results: FormatResult) -> int:
```

File: utils/automation/fossologyscanner.py [hunk 4]
```diff
@@ -245,15 +315,19 @@ def main(parsed_args):
   scanner = Scanners(cli_options)
   return_val = 0
 
+  # Populate tmp dir in unified diff format
+  format_results = FormatResult(cli_options)
+  format_results.process_files(scanner.cli_options.diff_dir)
+
```
Comment (GMishx) [inline utils/automation/fossologyscanner.py:321]:
```suggestion

```

File: utils/automation/fossologyscanner.py [hunk 5]
```diff
@@ -245,15 +315,19 @@ def main(parsed_args):
...
   result_dir = "results"
   os.makedirs(name=result_dir, exist_ok=True)
 
   if cli_options.report_format == ReportFormat.TEXT:
-    return_val = text_report(cli_options, result_dir, return_val, scanner)
+    return_val = text_report(cli_options, result_dir, return_val, scanner,format_results)
   else:
     return_val = bom_report(cli_options, result_dir, return_val, scanner,
-                            api_config)
+                            api_config,format_results)
```
Comment (GMishx) [inline utils/automation/fossologyscanner.py:330]:
```suggestion
                            api_config,format_results)
```

File: utils/automation/FoScanner/FormatResults.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+from .CliOptions import CliOptions
+
+class FormatResult:
+    """
+    For formatting the results from scanners with line number information
+    """
+
+    def __init__(self,cli_options:CliOptions):
+        self.cli_options = cli_options
```
Comment (GMishx) [inline utils/automation/FoScanner/FormatResults.py:19]: As a best practice, please define class variable `cli_options` in the class body as well.

File: utils/automation/FoScanner/FormatResults.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+      """
+      Format the diff content in a particular format with corrected line numbers.
+
+      :param: diff_content: str String to format
+      :return: str formatted_string
+      """
+      formatted_diff = []
+      diff_lines = diff_content.splitlines()
+      left = right = 0
+      ll = rl = 0
```
Comment (GMishx) [inline utils/automation/FoScanner/FormatResults.py:31]: Please use better names like `left_len` and `right_len`.

File: utils/automation/FoScanner/Scanners.py [hunk 8]
```diff
@@ -157,6 +157,62 @@ def get_copyright_list(self, all_results: bool = False) \
       return copyright_list
     return False
 
+  def get_copyright_whole(self, all_results: bool = False) \
```
Comment (GMishx) [inline utils/automation/FoScanner/Scanners.py:160]: The body of this function is very similar to `get_copyright_list()`. Any reason to define this new function?

**Review by GMishx — CHANGES_REQUESTED** [review]
Some questions and suggestion.

**Comment by GMishx** [comment]
While testing, I got the following error:

```
$ /bin/fossologyscanner copyright keyword
Traceback (most recent call last):
  File "/bin/fossologyscanner", line 347, in <module>
    sys.exit(main(args))
  File "/bin/fossologyscanner", line 320, in main
    format_results.process_files(scanner.cli_options.diff_dir)
  File "/bin/FoScanner/FormatResults.py", line 110, in process_files
    normal_string = file_contents.encode().decode('unicode_escape')
UnicodeDecodeError: 'unicodeescape' codec can't decode bytes in position 578-579: truncated \UXXXXXXXX escape
```

See: https://gitlab.com/GMishx/fossology/-/jobs/7210862637

File: utils/automation/fossologyscanner.py (hunk 1 above)
Comment (rajuljha) [reply utils/automation/fossologyscanner.py:86]: Okay! I'll keep that in mind!

File: utils/automation/fossologyscanner.py (hunk 3 above)
Comment (rajuljha) [reply utils/automation/fossologyscanner.py:207]: Yes. Corrected

File: utils/automation/FoScanner/FormatResults.py (hunk 6 above)
Comment (rajuljha) [reply utils/automation/FoScanner/FormatResults.py:19]: Okay! I've corrected it.

File: utils/automation/FoScanner/FormatResults.py (hunk 7 above)
Comment (rajuljha) [reply utils/automation/FoScanner/FormatResults.py:31]: I have corrected it in #2756

File: utils/automation/FoScanner/Scanners.py (hunk 8 above)
Comment (rajuljha) [reply utils/automation/FoScanner/Scanners.py:160]: get_copyright_list() only returned the copyright "content" from the keyword_scanner,
while get_copyright_whole() returns the "content", along with the byte info "start", "end" as well.

File: utils/automation/fossologyscanner.py (hunk 2 above)
Comment (rajuljha) [reply utils/automation/fossologyscanner.py:167]: Yes it was redundant. I forgot to remove it. I have corrected it

**Comment by rajuljha** [comment]
> While testing, I got the following error:
> 
> ```
> $ /bin/fossologyscanner copyright keyword
> Traceback (most recent call last):
>   File "/bin/fossologyscanner", line 347, in <module>
>     sys.exit(main(args))
>   File "/bin/fossologyscanner", line 320, in main
>     format_results.process_files(scanner.cli_options.diff_dir)
>   File "/bin/FoScanner/FormatResults.py", line 110, in process_files
>     normal_string = file_contents.encode().decode('unicode_escape')
> UnicodeDecodeError: 'unicodeescape' codec can't decode bytes in position 578-579: truncated \UXXXXXXXX escape
> ```
> 
> See: https://gitlab.com/GMishx/fossology/-/jobs/7210862637
> 
> Diff in question can be accessed from: https://gitlab.com/api/v4/projects/19657896/merge_requests/2/changes

* Does the files that we are scanning contain broken unicode escape seqeunces? 
* We can wrap it around with a try and except block to overcome this. Will the scan results be affected if the content contains unicode escape sequences? If not, we can skip this step as well.

File: utils/automation/FoScanner/Scanners.py (hunk 8 above)
Comment (GMishx) [reply utils/automation/FoScanner/Scanners.py:160]: Don't you think in such case it is better to extend the `get_copyright_list()` to return more and reduce the result as and when needed?

File: utils/automation/FoScanner/Scanners.py [hunk 9]
```diff
@@ -157,6 +157,61 @@ def get_copyright_list(self, all_results: bool = False) \
...
+        continue
+      if result['results'] is not None and result['results'] != "Unable to " \
+                                                                "read file":
+        contents = list()
+
+        for finding in result['results']:
+          if finding is not None and finding['type'] == "statement":
+            contents.append(finding)
+        if len(contents) > 0:
+          copyright_list.append(ScanResult(path, result['file'], contents))
```
Comment (GMishx) [inline utils/automation/FoScanner/Scanners.py:185]: `contents` here is breaking the datatype hint of 3rd parameter of `ScanResult`. It expects `set[str]` but getting `list[dict[str,str]]`

File: utils/automation/FoScanner/Scanners.py [hunk 10]
```diff
@@ -157,6 +157,61 @@ def get_copyright_list(self, all_results: bool = False) \
...
+          True:
+        continue
+      if result['results'] is not None and result['results'] != "Unable to " \
+                                                                "read file":
+        contents = list()
+        for finding in result['results']:
+          if finding is not None:
+            contents.append(finding)
+        if len(contents) > 0:
+          keyword_list.append(ScanResult(path, result['file'], contents))
```
Comment (GMishx) [inline utils/automation/FoScanner/Scanners.py:210]: Same datatype hint issue here.

File: utils/automation/FoScanner/Scanners.py (hunk 9 above)
Comment (GMishx) [reply utils/automation/FoScanner/Scanners.py:185]: Hints are put in place to suggest what kind of data I can expect here.

**Review by GMishx — CHANGES_REQUESTED** [review]

File: utils/automation/FoScanner/Scanners.py (hunk 8 above)
Comment (rajuljha) [reply utils/automation/FoScanner/Scanners.py:160]: Yeah! I see it now. For other function as well, their is a lot of unnecessary code duplication.
Can we maybe use a flag with the functions like `whole` which will default to false and we can pass it as true when we need the whole information from it.

File: utils/automation/FoScanner/Scanners.py (hunk 9 above)
Comment (rajuljha) [reply utils/automation/FoScanner/Scanners.py:185]: I'm on it. I'll correct it.

File: utils/automation/FoScanner/FormatResults.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+import os
+
+from .CliOptions import CliOptions
+
+class FormatResult:
+    """
+    For formatting the results from scanners with line number information
+
+    :ivar cli_options: CliOptions object
+    """
```
[the remark this answers is not in the document]
Comment (rajuljha) [reply utils/automation/FoScanner/FormatResults.py:17]: Okay. I got it now.

[the remark this answers is not in the document]
Comment (GMishx) [reply utils/automation/FoScanner/FormatResults.py:17]: @rajuljha , I cannot see the variable `cli_options` defined in the class as suggested in the comment above. Please avoid resolving comments as it makes reviewing the PR hard.

**Review by GMishx — CHANGES_REQUESTED** [review]

File: utils/automation/FoScanner/FormatResults.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (rajuljha) [reply utils/automation/FoScanner/FormatResults.py:17]: Okay! I will keep that in mind from now onwards.

**Review by GMishx — APPROVED** [review]

**Pull request merged**